garak/generators/azure.py

```diff
@@ -0,0 +1,140 @@
+    #. On the Azure portal page for the Azure OpenAI you want to use click on "Resource Management -> Keys and Endpoint" and copy paste the API Key and endpoint. 
+    #. In your console, Set the ``AZURE_API_KEY``, ``AZURE_ENDPOINT`` and ``AZURE_DEPLOYMENT_NAME`` variables.
+    #. Run garak, setting ``--model_type`` to ``azure`` and ``--model_name`` to [the name of the model](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models)
+    - e.g. ``gpt-4o``.
+    """
+
+    ENV_VAR = "AZURE_API_KEY"
+    active = True
+    generator_family_name = "Azure"
+    name_backup = ""
+
+    def __init__(self, name="", config_root=_config):
+        if name in chat_models_openai:
+            self.name = chat_models_openai[name]
+        elif name in completion_models_openai:
+            self.name = completion_models_openai[name]
+        else:
+            self.name = name
+
+        self._load_config(config_root)
+        if self.name in context_lengths:
+            self.context_len = context_lengths[self.name]
+
+        super().__init__(
+            self.name, generations=self.generations, config_root=config_root
+        )
+
+    def _load_client(self):
+        azure_endpoint = os.getenv("AZURE_ENDPOINT", default=None)
+        if azure_endpoint is None:
+            raise ValueError(
+                'The AZURE_ENDPOINT environment variable is required.\n'
+            )
+
+        azure_deployment_name = os.getenv("AZURE_DEPLOYMENT_NAME", default=None)
+        if azure_deployment_name is None:
+            raise ValueError(
+                'The AZURE_DEPLOYMENT_NAME environment variable is required.\n'
+            )
+
+        self.client = openai.AzureOpenAI(azure_endpoint=azure_endpoint, api_key=self.api_key)
+
+        if self.name == "":
+            openai_model_list = sorted([m.id for m in self.client.models.list().data])
+            raise ValueError(
+                f"Model name is required for {self.generator_family_name}, use --model_name\n"
+                + "  API returns following available models: ▶️   "
+                + "  ".join(openai_model_list)
+                + "\n"
+                + "  ⚠️  Not all these are text generation models"
+            )
+
+        if self.name in completion_models:
+            self.generator = self.client.completions
+        elif self.name in chat_models:
+            self.generator = self.client.chat.completions
+        elif "-".join(self.name.split("-")[:-1]) in chat_models and re.match(
+            r"^.+-[01][0-9][0-3][0-9]$", self.name
+        ):  # handle model names -MMDDish suffix
+            self.generator = self.client.completions
+        else:
+            raise ValueError(
+                f"No {self.generator_family_name} API defined for '{self.name}' in generators/openai.py - please add one!"
+            )
+
+        self.name_backup = self.name
+        self.name = azure_deployment_name
```

Comment:
Based on notes about validating env values once:
```suggestion
        self.name_backup = self.name
        self.name = self.azure_deployment_name
```
I am not super familiar with Azure OpenAI patterns, can you add comments around why `AZURE_DEPLOYMENT_NAME` would override name here or was this simply because the value is not currently held in the instance?

Reply:
The model name is not used while calling Azure OpenAPI, instead the deployment name should be used :
https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/switching-endpoints#keyword-argument-for-model

The deployment name could be anything as it is specified by the used.
The better choice could be to define the deployment name in command line parameter like that :
```
python -m garak --model_type azure --model_deployment_name mygpt --model_name gpt-4o --probes promptinject.HijackHateHumansMini
```
Otherwise having an environment variable is ok.